pull request #1

davsol opened the pull request:
Hi Joshua,
I reordered my version so that now I think it's finally good enough for merging..
Some of the major advances I've done:

Docker:
Added some verified Dockerfiles, for several versions of mxnet & cuda. 
** The latest Dockerfile is actually simpler, now that mxnet 0.12 is mostly compatible.. 

Training: 
* Saving ROC graphs
* training parameters - added optimizer choise, other small changes..
* net visualization
* Logging to tensorboard - 
         losses, 
         ROC timeline (ROC over time), 
         layer distribution

Dataset (Only pascal for now):
* create .rec for a sub-dataset of several classes - Clean dataset with only these classes' images
* Experimental: create dataset with "negative images" - meaning that none of the sub-classes is tagged in some of the images 

As i told you before, I will gladly contribute to this repo in the future, so that I'm looking forward to working together, and hope you will find my code usefull :)
Contact me via mail, for more debating and explainations..

Thanks, David
** solomond78@gmail.com

zhreshold commented on .idea/vcs.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Can you remove this file?

zhreshold commented on dataset/pascal_voc.py:
```diff
@@ -24,18 +24,22 @@ class PascalVoc(Imdb):
-        self.classes = self._load_class_names(names,
-            os.path.join(os.path.dirname(__file__), 'names'))
+        if class_names is not None:
+            self.classes = class_names.replace(' ', '').split(',')
```
class_names.strip().split(',') will be better IMO.

davsol replied:
I'll change that

zhreshold commented on dataset/pascal_voc.py:
```diff
@@ -156,8 +186,9 @@ def _load_image_labels(self):
                 if cls_name not in self.classes:
-                    continue
-                cls_id = self.classes.index(cls_name)
+                    cls_id = len(self.classes)
```
Do you intended to regard everything not in provided names as one class?

davsol replied:
Currently this is the best I could think of ..
at first I thought maybe to list them as ignores ('-1') but it was not a good hack

zhreshold commented on symbol/symbol_factory.py:
```diff
@@ -21,8 +60,7 @@ def get_config(network, data_shape, **kwargs):
             pads = [-1, -1, 1, 1, 1, 1, 1]
-            sizes = [[.07, .1025], [.15,.2121], [.3, .3674], [.45, .5196], [.6, .6708], \
-                [.75, .8216], [.9, .9721]]
+            sizes = get_scales(min_scale=0.15, max_scale=0.9, num_layers=len(from_layers))
```
Is this making slightly different anchor scales so the current models will perform worse than the original parameters?

davsol replied:
when you pick min_scale=0.2, max_scale=0.9 you get exactly the same numbers as you used. 
I now see that it is not correct for min_scale=0.15 
I guess my calculations were wrong. I'll check it out and get back to you

davsol replied:
Fixed, Fully compatible now..

zhreshold commented on symbol/symbol_factory.py:
```diff
@@ -33,7 +71,7 @@ def get_config(network, data_shape, **kwargs):
             strides = [-1, -1, 2, 2, 1, 1]
             pads = [-1, -1, 1, 1, 0, 0]
-            sizes = [[.1, .141], [.2,.272], [.37, .447], [.54, .619], [.71, .79], [.88, .961]]
+            sizes = get_scales(min_scale=0.2, max_scale=0.9, num_layers=len(from_layers))
```
same here.

davsol commented:
Sure :)

2017-11-06 21:38 GMT+02:00 Joshua Z. Zhang <notifications@github.com>:

> *@zhreshold* commented on this pull request.
>
> Awesome work! See comments for the reviews.
> Can you also add a tutorial section tell people how to use docker file and
> tensor board in README?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/zhreshold/mxnet-ssd/pull/134#pullrequestreview-74535486>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALzmeXqaxvV2esyL-JQS2v7tFxtJlXelks5sz2AtgaJpZM4QSARJ>
> .
>

zhreshold commented:
Please let me know if you are ready 😸

davsol commented:
I thought I'm the one who waits 😂 I'm ready..

2017-11-13 22:02 GMT+02:00 Joshua Z. Zhang <notifications@github.com>:

> Please let me know if you are ready 😸
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/zhreshold/mxnet-ssd/pull/134#issuecomment-344040312>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALzmeWW5e12epD6BJZO1we233mZWo5nuks5s2KBKgaJpZM4QSARJ>
> .
>

zhreshold commented:
Working on testing. Feel free to push more if you like.

davsol commented:
Great,
Is there a better way of communication ? I guess there will be several back
and forth, and currently each iteration takes us much time..
Can you easily see when I'm updating / commenting?

On Nov 21, 2017 8:28 PM, "Joshua Z. Zhang" <notifications@github.com> wrote:

> Working on testing. Feel free to push more if you like.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/zhreshold/mxnet-ssd/pull/134#issuecomment-346118160>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALzmeZjvOyM3KTaLqRCe8WSJ_nv7jWpPks5s4xZKgaJpZM4QSARJ>
> .
>

zhreshold commented:
@davsol I think it's okay to have conversations here. I am slowly pushing it because I am traveling these days, so bare with my latency. 😮

davsol commented:
Ok, I'm available for anything you need. Have a good time traveling :)

On Nov 22, 2017 8:30 AM, "Joshua Z. Zhang" <notifications@github.com> wrote:

> @davsol <https://github.com/davsol> I think it's okay to have
> conversations here. I am slowly pushing it because I am traveling these
> days, so bare with my latency. 😮
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zhreshold/mxnet-ssd/pull/134#issuecomment-346256199>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALzmeX5VTkVhIi8DxfYFa1jO-g14b--Jks5s47-cgaJpZM4QSARJ>
> .
>

zhreshold commented on train.py:
```diff
@@ -28,7 +29,7 @@ def parse_args():
                         default=1, type=int)
     parser.add_argument('--prefix', dest='prefix', help='new model prefix',
-                        default=os.path.join(os.getcwd(), 'model', 'ssd'), type=str)
+                        default=os.path.join(os.getcwd(), 'output', 'exp1', 'ssd'), type=str)
```
OSError: [Errno 2] No such file or directory: '/home/ubuntu/test/mxnet-ssd/output/exp1' if not explicitly created for this.

davsol replied:
I can erase that, or automatically create this folder.. what do you propose? 
I find it more convenient to have an 'output' folder for all experiments checkpoints, ROC graphs, detection vis and logs to tensorboard

zhreshold commented on symbol/symbol_factory.py:
```diff
@@ -1,7 +1,46 @@
-from __future__ import absolute_import
 import logging
-from . import symbol_builder
+import symbol_builder
```
this is broken in python3
```
from symbol import symbol_builder
```

zhreshold commented on symbol/symbol_factory.py:
```diff
@@ -1,7 +1,46 @@
+    step = int(np.floor((max_ratio - min_ratio) / (num_layers - 2)))
+    min_sizes = []
+    max_sizes = []
+    for ratio in xrange(min_ratio, max_ratio + 1, step):
```
xrange is for python2 only

davsol replied:
changing that to 'range' now

zhreshold commented:
Basically I've tested with python2, which is awesome, though I've met some tensorboard problems, that StringIO.StringIO bug is fixed in 1.0.0a7 but I can only find 1.0.0a6 on PYPI.
If I switch to python3 for testing, there are incompatible changes to fix.
Generally it's an awesome PR, but we need to get through by fixing these minor issues. @davsol

davsol commented:
Fixed what you've talked about, 
about python3, fixed what you've mentioned, but I can't really check if it works..

about tensorboard.. well, it's a bit complicated..
as I understand, there are copyright issues, tensorflow wants to have 'pip install tensorboard'. and so, the pip package is not maintained for now.
Since I've encountered the same problem as you did (StringIO.StringIO) I will update another README file, explaining this "must" code modification.
Anyway, if anyone would work with Docker, I've already handled this :)

Thanks for your kind words, I'm putting a lot of effort, making this repo the best @zhreshold

zhreshold commented:
@davsol Please fix the conflicts and we are good to merge now. Potential issues can be fixed later.

davsol commented:
I don't have write access so I can't solve this conflict.. it's a simple "if" case, that should not harm the original code..

davsol commented:
@zhreshold

zhreshold commented:
You can always fetch the upstream and fix the conflicts. Anyway, I am merging this, thanks for the contribution. @davsol
